### PR TITLE
Prepare GraphQL mutation in UI API Layer for creating K8s resources

### PR DESCRIPTION
**Pull request by tgorgol**
**Description**

Changes proposed in this pull request:

- Add ui-api-layer graphql mutation for creating resources from JSON 
- Add ui-api-layer acceptance tests for said mutation
- Modify ui-api-layer authorization to allow extracting resource meta from JSON passed in argument
- Add new error type and slightly modify existing gqlerror package
- Add troubleshooting section to ui-api-layer README file, with information about recently encountered authorization issue

**Related issue(s)**

#2776 #2971

### CODE REVIEW

File: components/ui-api-layer/README.md [hunk 1]
```diff
@@ -87,3 +87,12 @@ go test ./...
...
 To check if the code is correct and you can push it, run the `before-commit.sh` script. It builds the application, runs tests, checks the status of the vendored libraries, runs the static code analysis, and ensures that the formatting of the code is correct.
+
+## Troubleshooting
+
+### Certificate signed by unknown authority
+```bash
+oidc.go:222] oidc authenticator: initializing plugin: Get https://dex.kyma.local/.well-known/openid-configuration: x509: certificate signed by unknown authority
+```
+This error might be related to issue with Go versions before 1.11.6 on macOS, discussed [here](https://github.com/golang/go/issues/24652).
+
```
Comment (kazydek) [inline components/ui-api-layer/README.md]: I would remove the Troubleshooting section and add it as Heading 3 under the Usahe section ( at the end)

```

### Certificate signed by an unknown authority

When you build the UI API Layer project, you can get the following error:

```bash
oidc.go:222] oidc authenticator: initializing plugin: Get https://dex.kyma.local/.well-known/openid-configuration: x509: certificate signed by unknown authority
```

This error can occur if you use Go version 1.11.5 on macOS. Try upgrading to version 1.11.6 or higher. For details, see [this](https://github.com/golang/go/issues/24652) issue.

```

File: components/ui-api-layer/README.md [hunk 2]
```diff
@@ -87,3 +87,12 @@ go test ./...
 ### Verify the code
 
 To check if the code is correct and you can push it, run the `before-commit.sh` script. It builds the application, runs tests, checks the status of the vendored libraries, runs the static code analysis, and ensures that the formatting of the code is correct.
+
+## Troubleshooting
```
Comment (kazydek) [inline components/ui-api-layer/README.md]:
```suggestion
### Certificate signed by an unknown authority

When you build the UI API Layer project, you can get the following error:

```bash
oidc.go:222] oidc authenticator: initializing plugin: Get https://dex.kyma.local/.well-known/openid-configuration: x509: certificate signed by unknown authority
```

This error can occur if you use Go version 1.11.5 on macOS. Try upgrading to version 1.11.6 or higher. For details, see [this](https://github.com/golang/go/issues/24652) issue.
```

File: components/console-backend-service/internal/domain/k8s/resource_service.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	client discovery.DiscoveryInterface
+}
+
+func newResourceService(client discovery.DiscoveryInterface) *resourceService {
+	return &resourceService{
+		client: client,
+	}
+}
+
+func (svc *resourceService) Create(namespace string, resource types.Resource) (types.Resource, error) {
```
Comment (michal-hudy) [inline components/console-backend-service/internal/domain/k8s/resource_service.go]: To be consistent with other services, return pointer.

File: resources/core/charts/cluster-users/templates/rbac-roles.yaml [hunk 4]
```diff
@@ -82,7 +82,7 @@ metadata:
   annotations:
     "helm.sh/hook-weight": "0"
 rules:
-- apiGroups: ["", "apps", "authentication.kyma-project.io", "extensions", "gateway.kyma-project.io", "kubeless.io", "rbac.authorization.k8s.io", "applicationconnector.kyma-project.io", "servicecatalog.k8s.io", "servicecatalog.kyma-project.io", "settings.k8s.io", "kyma.cx", "authentication.istio.io", "config.istio.io", "eventing.kyma-project.io", "ui.kyma.cx", "ui.kyma-project.io", "metrics.k8s.io", "networking.istio.io", "batch", "extensions", "autoscaling"]
+- apiGroups: ["", "apps", "authentication.kyma-project.io", "extensions", "gateway.kyma-project.io", "kubeless.io", "rbac.authorization.k8s.io", "applicationconnector.kyma-project.io", "servicecatalog.k8s.io", "servicecatalog.kyma-project.io", "settings.k8s.io", "kyma.cx", "authentication.istio.io", "config.istio.io", "eventing.kyma-project.io", "ui.kyma.cx", "ui.kyma-project.io", "metrics.k8s.io", "networking.istio.io", "batch", "extensions", "autoscaling", "assetstore.kyma-project.io"]
```
Comment (michal-hudy) [inline resources/core/charts/cluster-users/templates/rbac-roles.yaml]: Add also `cms.kyma-project.io` 😄

File: components/console-backend-service/internal/authz/auth.go [hunk 5]
```diff
@@ -40,11 +44,31 @@ func NewAuthorizer(client authorizationclient.SubjectAccessReviewInterface, cach
 }
 
 // PrepareAttributes prepares attributes for authorization
-func PrepareAttributes(ctx context.Context, u user.Info, attributes gqlschema.ResourceAttributes, obj interface{}) (authorizer.Attributes, error) {
+func PrepareAttributes(ctx context.Context, u user.Info, attributes gqlschema.ResourceAttributes, obj interface{}, client discovery.DiscoveryInterface) (authorizer.Attributes, error) {
```
Comment (mszostok) [inline components/console-backend-service/internal/authz/auth.go]: This method has really high cyclomatic complexities.
gocyclo calculated the complexity as **34** points wherein `internal` pkg we have avg 3pkt.

This method has 122 lines. Has a lot of `if, else, switch` which are even nested, check that:

```go
	if attributes.IsChildResolver {
		for i := 0; i < val.NumField(); i++ {
			if attributes.NameArg != nil {
				if nameArg := *attributes.NameArg; fieldName == nameArg {
					if val.Field(i).Kind() != reflect.String {
					}
				}
			}
			if attributes.NamespaceArg != nil {
				if namespaceArg := *attributes.NamespaceArg; fieldName == namespaceArg {
					if val.Field(i).Kind() != reflect.String {
					}
				}
			}
		}
		if attributes.NameArg != nil && name == "" {
		}
		if attributes.NamespaceArg != nil && namespace == "" {
		}
	} else {
```

in one `if` statement u have `for` in which u have `if` in which u have `if` in which u have `if`.

You should avoid to write method like that. You should extract some logic to dedicated function sand the name of the function also needs to be self-describing :)

File: components/console-backend-service/internal/authz/auth_test.go [hunk 6]
```diff
@@ -87,7 +90,37 @@ func TestPrepareAttributes(t *testing.T) {
...
+					assert.Error(t, err)
+				})
+				t.Run("Then authAttributes is empty", func(t *testing.T) {
+					assert.Nil(t, authAttributes)
+				})
+			})
+		})
+	})
+
+	t.Run("When resource meta information are required", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (tgorgol) [reply components/console-backend-service/internal/authz/auth_test.go]: I agree, however I wanted to be consistent with how the tests in that file were before my contribution.

File: components/console-backend-service/internal/authz/auth.go (hunk 5 above)
Comment (tgorgol) [reply components/console-backend-service/internal/authz/auth.go]: Great points, some of your comments were not related to my contribution (`if in if in if in for in if`), but I totally agree that the function requires a refactor. Never heard about gocyclo, so I'm glad you mentioned it. :)

**Pull request merged**